[Inductor] [FX passes] Group linear fusion

yanboliang opened the pull request:
Summary:
The draft version of a group + batch fusion framework, and the group linear fusion implementation.
In the future, it's pretty straightforward to add a new group/batch fusion policy by defining a class with match + fuse functions.

Test Plan: buck2 test 'fbcode//mode/dev-nosan' fbcode//caffe2/test/inductor:group_batch_fusion

Differential Revision: D46956695



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

yanboliang commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -0,0 +1,183 @@
+            and len(node.args[1].meta["tensor_meta"].shape) == 2
+            and node.kwargs.get("beta", 1.0) == 1.0
+            and node.kwargs.get("alpha", 1.0) == 1.0
+        ):
```
```aten.mm``` can be fused here as well, I'll add it after fbgemm team relaxes the limitation that group transpose should be consistent.

yanboliang commented on torch/_inductor/config.py:
```diff
@@ -48,6 +48,9 @@
 split_cat_fx_passes = True
 
+# enable pattern match with group fusion (using fbgemm)
+group_fusion = False
```
The major reason of making it false is: fbgemm group linear uses a different linear algebra compared to PT, so it's not numerical on par with PT linear, differing by the 4th digit right now.

yanboliang replied:
Unfortunately it's worse compared to fp64. I'm testing with [```same```](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/utils.py#L911) function, and this is the output:
```
[2023-07-13 16:16:05,949] torch._dynamo.utils: [ERROR] RMSE (res-fp64): 0.00017, (ref-fp64): 0.00000 and shape=torch.Size([4, 16])
[2023-07-13 16:16:06,108] torch._dynamo.utils: [ERROR] RMSE (res-fp64): 0.00004, (ref-fp64): 0.00000 and shape=torch.Size([4, 6])
```

> Can you do a full CI run with this enabled?

The unit tests are running with ```group_fusion ``` enabled, but I think OSS CI doesn't have ```fbgemm```, so they are skipped. In fbcode, I added ```fbgemm``` as inductor unit tests dependency, so they are tested. Do you think we should set up ```fbgemm``` for OSS CI as well?

jansel replied:
We should open source the grouped gemm code.

yanboliang commented:
I’ll do a full CI run with it enabled and update result later.

On Thu, Jul 13, 2023 at 1:02 PM Jason Ansel ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In torch/_inductor/config.py
> <https://github.com/pytorch/pytorch/pull/105116#discussion_r1263000941>:
>
> > @@ -48,6 +48,9 @@
>  # Optimize away split cat patterns (Experimental)
>  split_cat_fx_passes = True
>
> +# enable pattern match with group fusion (using fbgemm)
> +group_fusion = False
>
> Is it better or worse compared to float64?
>
> Can you do a full CI run with this enabled?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/105116#discussion_r1263000941>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAO7AKUBDVJ6WF3RSFX4PQTXQBH63ANCNFSM6AAAAAA2IMRRDU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jansel reviewed (approved):
Approving to unblock, we should open source the grouped gemm code.

yanboliang commented:
@jansel Thanks for your review, sharing our plan for the group/batch fusion:
* I'll land this PR to unblock internal usage and other group/batch fusion development. ```group_fusion``` flag will be False by default.
* Batch fusion uses this framework as well, but they don't depend on ```fbgemm```, so we will have ```batch_fusion``` flag True by default.
* I've checked with ```fbgemm``` team, there is no obvious blocker to open source ```gmm``` related code, only bandwidth limitation. I'll push them to open source it.
* After ```gmm``` is open source, I'll work with PyTorch Dev Infra to set up CI with ```gmm``` and make sure all gmm related changes can be tested in OSS CI as well. And we will evaluate if we should enable ```group_fusion``` by default.

Let me know if you have any suggestion!

yanboliang committed:
[Inductor] [FX passes] Group linear fusion (#105116)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/105116

The draft version of a group + batch fusion framework, and the group linear fusion implementation.
In the future, it's pretty straightforward to add a new group/batch fusion policy by defining a class with match + fuse functions.

Test Plan: buck2 test 'fbcode//mode/dev-nosan' fbcode//caffe2/test/inductor:group_batch_fusion

Reviewed By: jackiexu1992

Differential Revision: D46956695

fbshipit-source-id: 47677b9b66c22597272410275803f96a239ce437

jansel commented:
> @jansel Thanks for your review, sharing our plan for the group/batch fusion:
> 
>     * I'll land this PR to unblock internal usage and other group/batch fusion development. `group_fusion` flag will be False by default.
> 
>     * Batch fusion uses this framework as well, but they don't depend on `fbgemm`, so we will have `batch_fusion` flag True by default.
> 
>     * I've checked with `fbgemm` team, there is no obvious blocker to open source `gmm` related code, only bandwidth limitation. I'll push them to open source it.
> 
>     * After `gmm` is open source, I'll work with PyTorch Dev Infra to set up CI with `gmm` and make sure all gmm related changes can be tested in OSS CI as well. And we will evaluate if we should enable `group_fusion` by default.
> 
> 
> Let me know if you have any suggestion!

Sounds good.  We might also be able to just copy it into PyTorch.  I believe it is pretty thin wrapper around CUTLASS.

yanboliang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1395,6 +1395,7 @@ def can_fallback(func: OpOverload):
                 "vision",
                 "torchtext",
                 "torchaudio",
+                "fbgemm",
```
Resolved by offline discussion with @zou3519. #106210 is a temporary fix and I'll add meta registration for fbgemm.gmm later.
